Corona-Warn-App/src/main/java/de/rki/coronawarnapp/risk/DefaultRiskLevels.kt

```diff
@@ -207,14 +230,278 @@ class DefaultRiskLevels @Inject constructor(
+        exposureWindow.scanInstances.fold(.0) { seconds, scanInstance ->
+            val weight =
+                minutesAtAttenuationWeight
+                    .filter { it.attenuationRange.inRange(scanInstance.typicalAttenuationDb) }
+                    .map { it.weight }
+                    .firstOrNull() ?: .0
+            return seconds + scanInstance.secondsSinceLastScan * weight
+        }
+
+    private fun determineRiskLevel(
+        normalizedTime: Double,
+        timeToRiskLevelMapping: List<RiskCalculationParametersOuterClass.NormalizedTimeToRiskLevelMapping>
+    ) =
+        timeToRiskLevelMapping
+            .filter { it.normalizedTimeRange.inRange(normalizedTime) }
+            .map { it.riskLevel }
+            .firstOrNull()
+
+    override suspend fun calculateRisk(
+        exposureWindow: ExposureWindow
+    ): RiskResult? {
+        if (dropDueToMinutesAtAttenuation(exposureWindow, appConfig.minutesAtAttenuationFilters)) {
+            return null
+        }
+
+        val transmissionRiskLevel = determineTransmissionRiskLevel(
+            exposureWindow,
+            appConfig.transmissionRiskLevelEncoding
+        )
+
+        if (dropDueToTransmissionRiskLevel(transmissionRiskLevel, appConfig.transmissionRiskLevelFilters)) {
+            return null
+        }
+
+        val transmissionRiskValue =
+            transmissionRiskLevel * appConfig.transmissionRiskLevelMultiplier
+
+        val weightedMinutes = determineWeightedSeconds(
+            exposureWindow,
+            appConfig.minutesAtAttenuationWeights
+        ) / 60
+
+        val normalizedTime = transmissionRiskValue * weightedMinutes
+
+        val riskLevel = determineRiskLevel(
+            normalizedTime,
+            appConfig.normalizedTimePerExposureWindowToRiskLevelMapping
+        )
+            ?: throw NormalizedTimePerExposureWindowToRiskLevelMappingMissingException()
+
+        return RiskResult(transmissionRiskLevel, normalizedTime, riskLevel)
+    }
+
+    override suspend fun aggregateResults(
+        exposureWindowsAndResult: Map<ExposureWindow, RiskResult>
+    ): AggregatedRiskResult {
+        val uniqueDatesMillisSinceEpoch = exposureWindowsAndResult.keys
+            .map { it.dateMillisSinceEpoch }
+            .toSet()
+
+        Timber.d(
+            "uniqueDates: ${
+                TextUtils.join(System.lineSeparator(), uniqueDatesMillisSinceEpoch)
+            }"
+        )
+        val exposureHistory = uniqueDatesMillisSinceEpoch.map {
+            aggregateRiskPerDate(it, exposureWindowsAndResult)
+        }
+
+        exposureHistory.forEach {
+            Timber.d(
+                "(date=${it.dateMillisSinceEpoch}, " +
+                    "riskLevel=${it.riskLevel}, " +
+                    "minimumDistinctEncountersWithLowRisk=${it.minimumDistinctEncountersWithLowRisk}, " +
+                    "minimumDistinctEncountersWithHighRisk=${it.minimumDistinctEncountersWithHighRisk})"
+            )
+        }
```

Comment:
forEach is always executed, even if Timber logging is off. I'm sure there's some API to check if debug level logging is on, right? Might be worth wrapping this block with an if-statement.

Reply:
thats true.. It is also redundant logging, because major infos are logged already in the steps before. 
I cleaned this. thx